### DISCUSSION TITLE
Initialize (begin) instance of SX1278 without reset (after deep sleep)

### DISCUSSION THREAD

**Discussion by sacculus** [jgromes/RadioLib#519] [Q&A]
Hi
Is it any procedure to initialise instance of SX1278 without reset LoRa chip?
I'd try to put MCU (esp32) in deep sleep when LoRa modem (sx1278) in receive state.
When packet received MCU is woken up by extX wake up source with dio0 line. But begin() method of instance SX1278 resets LoRa chip and its state lost, so cant read received packet, its size, RSSI etc. Without initialize I can't read resieved data too, because there is not control SS line of SPI.
Temporary I commented out code with reset in SX127x::findChip routine but sure that is not the best decision to change library code...
Is it any right way to do this?
I think I am not first who plays with power saving of LoRa boards.

Thank you!

**Comment by jgromes** [comment]
The reset in begin method is quite intentional - the assumption is that you're only calling begin() after power-on reset, when the SX127x device is in an unknown state.

However, I'm not sure why would you want to call begin() on wakeup of the ESP32 in the first place? I would only call begin once, after the power-on reset.

EDIT: Also, this doesn't seem like an issue - converting to a discussion instead.

**Comment by DarkZeros** [comment]
I am reviving this discussion since i have exactly same issue. I will try to fix it myself and propose a PR.

 The issue is that ESP32 can set EXT1 /EXT0 wakeup to IRQ (DIO1) and enter deep sleep.

This saves huge amounts of power, but when a packet is received, the module state needs to be restored without the initialization (memory is fully reset expect RTC memory). And the SX126x does not know about the ESP32 being in deep sleep.

**Reply by StevenCellist** [reply]
Depending on how power constrained you are, you may be able to survive on light-sleep. On most boards, lightsleep can go down to 2mA or a bit lower, which compared to the radio staying in Rx (12mA) isn't bad at all. Plus it saves all the hassle of re-initializing everything.

**Reply by DarkZeros** [reply]
I am working on a solar powered project, therefore every uA I can save is welcomed. 

The ESP32 itself uses 1mA in light sleep and 0.1uA in deep sleep. 
The radio uses 4.5mA in receiving state. Therefore it is a 25% power save.

Technically it is possible to save/restore the key members of the Module class, into RTC memory, or flag the fields accordingly. I will keep you posted on the progress.

**Reply by StevenCellist** [reply]
You are replying on a topic that mentions SX1278 which has a reported Rx power consumption of 12mA. 4.5mA sounds suspiciously like an SX1262 instead. So that is something to watch out for.

Moreover, why not use an STM32? It is absurdly more power-efficient nor does it lose its RAM contents. No need to do any sort of RTC hackery with possibly very tough to solve bugs.

**Reply by DarkZeros** [reply]
Yes, it is a SX1262.
Unfortunately i am too advance in the project now to change the main MCU for a STM32.

But so far the esp32 has been great, managing to run on average with just 5uA (as long as i keep the deep sleep as much as i can, around 99.5% of the time)

**Comment by HeadBoffin** [comment]
>  I will try to fix it myself and propose a PR.

As @jgromes said in 2022, this isn't an issue or a bug so there isn't anything really to fix - you can get the reason for wake from the ESP32, if it's wake from sleep because of external interrupt, just don't call radio.begin() and proceed directly to reading the packet from the radio - it really is that simple.

**Reply by DarkZeros** [reply]
In theory yes, it is simple, but there are parts of the initialization that need to happen in begin(). Otherwise the module is not in a valid state to read data from it. I tried and got this result using the read() on an uninitialized SX1262.

> E (31) lora: wakeup
> [     4][E][esp32-hal-gpio.c:190] __digitalRead(): IO 36 is not set as GPIO.
> [    16][E][esp32-hal-gpio.c:176] __digitalWrite(): IO 35 is not set as GPIO.
> [    23][E][esp32-hal-gpio.c:176] __digitalWrite(): IO 35 is not set as GPIO.
> [    30][E][esp32-hal-gpio.c:190] __digitalRead(): IO 36 is not set as GPIO.
> ...
> [   204][E][esp32-hal-gpio.c:190] __digitalRead(): IO 36 is not set as GPIO.
> [   211][E][esp32-hal-gpio.c:176] __digitalWrite(): IO 35 is not set as GPIO.
> [   218][E][esp32-hal-gpio.c:176] __digitalWrite(): IO 35 is not set as GPIO.
> [   225][E][esp32-hal-gpio.c:190] __digitalRead(): IO 36 is not set as GPIO.
> E (259) lora: data: ��������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������

**Reply by HeadBoffin** [reply]
Hmmmm, the Arduino core general initialises GPIO to the appropriate mode on each call - which is why it is so slow.

Are you using ESP-IDF?

**Reply by StevenCellist** [reply]
With IDF v5 (which is what the Arduino core v3.x runs on) this is not the case anymore. Plus, the RadioLib drivers aren't completely stateless and do use a bunch of variables that would be lost upon deepsleep wakeup.

**Reply by HeadBoffin** [reply]
The GPIO setup code is in the Arduino core of all other implementations and the only bit that doesn't appear to be working from above is the setup of the GPIO. Perhaps simpler to wait for the OP to confirm their environment.

**Reply by DarkZeros** [reply]
Yes, I am using IDF v5.4 + Arduino v3 as a component. Arduino used to setup GPIOs before in v2 but it does not in the latest version.
Manually re-adding the GPIO pins does solve those issues, so they are actually minor.
However, even with that, the receive command is failing, storing the RadioLib objects in RTC mem does not solve the problem, I tried with same results...


The current code I have is this one:
```
const SPISettings Lora::kSpiSettings{2'000'000, MSBFIRST, SPI_MODE0};
RTC_DATA_ATTR ArduinoHal Lora::kHal(SPI, Lora::kSpiSettings);
RTC_DATA_ATTR Module Lora::kModule(&Lora::kHal, HW::Lora::Cs, HW::Lora::Dio1, HW::Lora::Rst, HW::Lora::Busy);
RTC_DATA_ATTR SX1262 Lora::kRadio(&Lora::kModule);
```

I will omit the EXT1 boot and other minor irrelevant code and just jump to a normal receive vs the broken receive call.
Normal receive:

> [9.125417] RLB_SPI: CMDR        11
> [9.125536] RLB_SPI: SI          0       0
> [9.125613] RLB_SPI: SO  D4      D4      1
> [9.125786] RLB_SPI: CMDR        13
> [9.125856] RLB_SPI: SI          0       0       0
> [9.127697] RLB_SPI: SO  D4      D4      F       0
> [9.130819] RLB_SPI: CMDR        C0
> [9.133339] RLB_SPI: SI          0       0
> [9.135943] RLB_SPI: SO  D4      54      54
> [9.138978] RLB_SPI: CMDR        12
> [9.141498] RLB_SPI: SI          0       0       0
> [9.144275] RLB_SPI: SO  D4      D4      0       12
> [9.147478] RLB_SPI: CMDR        11
> [9.150004] RLB_SPI: SI          0       0
> [9.152608] RLB_SPI: SO  D4      D4      1
> [9.155554] RLB_SPI: CMDR        13
> [9.158076] RLB_SPI: SI          0       0       0
> [9.160854] RLB_SPI: SO  D4      D4      F       0
> [9.164065] RLB_SPI: CMDR        1E      0
> [9.166669] RLB_SPI: SI                  0       0       0       0       0       0       0      00       0       0       0       0       0       0       0
> [9.171790] RLB_SPI: SO  D4      D4      D4      48      65      6C      6C      6F      20     57       6F      72      6C      64      21      20      23      30
> [9.178720] RLB_SPI: CMDW        2
> [9.181164] RLB_SPI: SI          43      FF
> [9.183942] RLB_SPI: SO  D4      D4      D4
> E (9179) Radio: received: Hello World! #0

Broken receive after deep sleep wakeup:
> E (31) lora: wakeup
> [0.005124] RLB_SPI: CMDR
> [0.010313] RLB_SPI: SI  0       0
> [0.010403] RLB_SPI: SO  D4      D4
> [0.010556] RLB_SPI: CMDR
> [0.010614] RLB_SPI: SI  0       0       0
> [0.011619] RLB_SPI: SO  D4      D4      D4
> [0.014650] RLB_SPI: CMDR
> [0.016914] RLB_SPI: SI  0       0
> [0.019431] RLB_SPI: SO  D4      D4
> [0.022198] RLB_SPI: CMDR
> [0.024465] RLB_SPI: SI  0       0       0
> [0.027156] RLB_SPI: SO  D4      D4      D4
> [0.030177] RLB_SPI: CMDR
> [0.032450] RLB_SPI: SI  0       0
> [0.034967] RLB_SPI: SO  D4      D4
> [0.037732] RLB_SPI: CMDR
> [0.040002] RLB_SPI: SI  0       0       0
> [0.042692] RLB_SPI: SO  D4      D4      D4
> [0.047116] RLB_SPI: CMDR        1E      D4
> [0.048508] RLB_SPI: SI                  0       0       0       0       0       0       0      00       0       0       0       0       0       0       0       0       0       0       0      00       0       0       0       0       0       0       0       0       0       0       0      00       0       0       0       0       0       0       0       0       0       0       0      00       0       0       0       0       0       0       0       0       0       0       0      00       0       0       0       0       0       0       0       0       0       0       0      00       0       0       0       0       0       0       0       0       0       0       0      00       0       0       0       0       0       0       0       0       0       0       0      00       0       0       0       0       0       0       0       0       0       0       0      00       0       0       0       0       0       0       0       0       0       0       0      00       0       0       0       0       0       0       0       0       0       0       0      00       0       0       0       0       0       0       0       0       0       0       0      00       0       0       0       0       0       0       0       0       0       0       0      00       0       0       0       0       0       0       0       0       0       0       0      00       0       0       0       0       0       0       0       0       0       0       0      00       0       0       0       0       0       0       0       0       0       0       0      00       0       0       0       0       0       0       0       0       0
> [0.087827] RLB_SPI: SO  D4      D4      D4      CC      47      9E      D3      B5      ED     D5       38      28      1C      7A      EC      3A      F6      D0      76      D4      3C     58       16      60      8A      16      12      AC      69      82      A3      5E      D2     DB       A7      E3      A8      ED      7A      53      84      15      C2      32      C7     74       53      48      65      6C      6C      6F      20      57      6F      72      6C     64       21      20      23      30      5F      E1      C0      AB      93      E0      7E     B6       6B      F4      A7      C9      77      E0      C4      93      73      71      6D     C9       77      B8      55      AF      68      BA      9C      9B      3       D4      47     2A       C4      E0      A7      A1      25      1E      55      75      8A      3A      46     E0       69      81      68      67      1       FB      C4      22      C7      E5      E9     A1       33      BD      34      21      B4      6F      F       B5      5B      D1      6D     63       78      71      36      64      79      64      2F      CB      B7      21      27     77       14      BA      F2      49      55      7B      93      BA      3E      3E      3      C1       BB      60      DF      1C      86      FE      F2      8F      F5      7F      8B     31       E5      2       82      8B      3B      DB      4B      74      D2      95      52     D7       9E      36      A0      EC      3A      BB      9A      3       CF      E6      10     89       2F      6D      66      7B      4B      EF      58      E6      2B      9E      C3     30       B       D3      3F      96      ED      87      FA      52      F1      C6      AB     C8       F7
> [0.145446] RLB_SPI: CMDW
> [0.147717] RLB_SPI: SI  43      FF
> [0.150408] RLB_SPI: SO  D4      D4
> E (179) lora: data: �G�ӵ��8(z�:��v�<X`��i��^�����zS��2�tSHello World! #0_�����~�k���w�ēsqm�w�U�h����G*�ৡ%Uu�:F�i�hg��"����3�4!�o�[�mcxq6dyd/˷!'w��IU{��>>��`������1���;�KtҕRמ6��:�����/mf{K�X�+��0
>   �?���R�ƫ��

**Reply by DarkZeros** [reply]
Update: I manage to make it work if I call:
```
void SX126x::modSetupBasic() {
  // set module properties
  this->mod->init();
  this->mod->hal->pinMode(this->mod->getIrq(), this->mod->hal->GpioModeInput);
  this->mod->hal->pinMode(this->mod->getGpio(), this->mod->hal->GpioModeInput);
  this->mod->spiConfig.widths[RADIOLIB_MODULE_SPI_WIDTH_ADDR] = Module::BITS_16;
  this->mod->spiConfig.widths[RADIOLIB_MODULE_SPI_WIDTH_CMD] = Module::BITS_8;
  this->mod->spiConfig.statusPos = 1;
  this->mod->spiConfig.cmds[RADIOLIB_MODULE_SPI_COMMAND_READ] = RADIOLIB_SX126X_CMD_READ_REGISTER;
  this->mod->spiConfig.cmds[RADIOLIB_MODULE_SPI_COMMAND_WRITE] = RADIOLIB_SX126X_CMD_WRITE_REGISTER;
  this->mod->spiConfig.cmds[RADIOLIB_MODULE_SPI_COMMAND_NOP] = RADIOLIB_SX126X_CMD_NOP;
  this->mod->spiConfig.cmds[RADIOLIB_MODULE_SPI_COMMAND_STATUS] = RADIOLIB_SX126X_CMD_GET_STATUS;
  this->mod->spiConfig.stream = true;
  this->mod->spiConfig.parseStatusCb = SPIparseStatus;
}
```
Instead of begin(). I will keep working on a nicer solution, and if it is neat and modular, push a PR.
I am also open to ideas to implement.

**Reply by StevenCellist** [reply]
Exactly why are you creating a custom ArduinoHAL, as well as a Module, as well as a Radio? Where did you get that from? The examples here definitely don't show it. No need for any of that - just check [the wiki](https://github.com/jgromes/RadioLib/wiki/Basics#non-standard-spi-setup) on how to setup your radio with a custom SPI instance. Way smaller chance of weird stuff happening! (Link from the FAQ on the wiki.)

**Reply by DarkZeros** [reply]
> Exactly why are you creating a custom ArduinoHAL, as well as a Module, as well as a Radio? Where did you get that from? The examples here definitely don't show it. No need for any of that - just check [the wiki](https://github.com/jgromes/RadioLib/wiki/Basics#non-standard-spi-setup) on how to setup your radio with a custom SPI instance. Way smaller chance of weird stuff happening! (Link from the FAQ on the wiki.)

I did it in an attempt to keep all the object's memory in RTC (persistent), however I failed, since it doesn´t work and fails the same way as just:
`SX1262 Lora::kRadio = new Module(HW::Lora::Cs, HW::Lora::Dio1, HW::Lora::Rst, HW::Lora::Busy, SPI, Lora::kSpiSettings);`

**Reply by HeadBoffin** [reply]
Sounds like progress, once it's working I'm sure some collective input on tidy up can be done.

**Reply by jgromes** [reply]
@DarkZeros I think the fundamental problem you will run into is that it is not possible to retrieve the SX1262 state from the device itself. For example, when we change the spreading factor, we send a command to the SX1262 device. However, there is no command to retrieve the currently configured spreading factor. That's why RadioLib needs a whole bunch of state variables in RAM (intialized on startup) to keep track of the device state.

ESP32 loses RAM in deep sleep, so the state of the SX1262 device is unknown after waking up, and the only way to get it into a known state again is running begin again. Even if you manage to make it working by re-initializing the ESP32 side of the connection (which is what `SX126x::modSetupBasic` is essentially doing), you will run into issues with the library thinking the device is in state A, but in actuality it's in state B. The only (hacky!) workaround I can think of is to run `SX126x::modSetupBasic` to get SPI communication working, then reading the packet contents (if there is one) and then doing a full reinit with begin().

**Reply by DarkZeros** [reply]
Hi, I managed to avoid the recreation of the objects, and keep everything in RTC RAM!

```
RTC_DATA_ATTR std::optional<SPISettings> kSpi;
RTC_DATA_ATTR std::optional<ArduinoHal> kHal;
RTC_DATA_ATTR std::optional<Module> kModule;
RTC_DATA_ATTR std::optional<SX1262> kRadio;

void setup() {
  if (!kRadio) {
    ESP_LOGE("lora", "initialize");
    // Construct the RadioLib objects just once in RTC mem
    kSpi.emplace(2'000'000, MSBFIRST, SPI_MODE0);
    kHal.emplace(SPI, *kSpi);
    kModule.emplace(&kHal.value(), HW::Lora::Cs, HW::Lora::Dio1, HW::Lora::Res, HW::Lora::Busy);
    kRadio.emplace(&kModule.value());
    kRadio->begin(434.0, 125.0, 9, 7, 0x12, 10, 8, 0, false);
    kRadio->startReceive();
  }
  
  // Perform basic initialization, set back the GPIO and IRQ
  kHal->pinMode(kModule->getIrq(), kHal->GpioModeInput);
  kHal->pinMode(kModule->getGpio(), kHal->GpioModeInput);
  kModule->init();
}
```

The output shows that during deep sleeps the constructors are not called, and data is received properly calling read():

> E (64) lora: initialize
> E (65) : SX1262 contructor
> E (1979) lora: listening
> E (31) lora: wakeup
> E (32) lora: data: Hello World! #0
> E (31) lora: wakeup
> E (32) lora: data: Hello World! #1

The trick resides on the std::optional<>, without it, the constructor will always be called on the object memory. 
Using std::optional properly wraps the memory for RTC, and survives across deep sleep, and retains the object's values.
I still need to manually call init() on the module and set the 2 pin modes of the pins that are normally done in modSetup().  Since those are lost, Luckily having access to the underlying objects from outside means I can do that without modifying the RadioLib at all.



My question is, i this safe? Will I run into any issues with it?
Is the IRQ handling properly setup this way or am I missing something? Do I need to re-set the Interrup handling or is it fine if I just leave the IRQ uninitialized and go back to sleep?

After aproximately 100-200 packets my lora device stops receiving, so I am not sure if there is something I broke configuring it this way. Or it is just that I have a bug somewhere else in my code.

**Reply by Griffon1412** [reply]
Hi all.

first of all I have not so much experience concerning Semtech's LoRa chips and I appreciate @jgromes  time and effort spent on RadioLib.
But may I asked why it is so complicated, maybe even impossible to use ESP32's deep sleep feature in combination with RadioLib? At least with SX126x chips?

I stumbled across another library for SX126x chips, [https://github.com/beegee-tokyo/SX126x-Arduino](https://github.com/beegee-tokyo/SX126x-Arduino).
It seems to have no problems at all with deep sleep. It's deep sleep example runs flawless on my Heltec and LilyGo LoRa boards.
There's a function lora_hardware_re_init() to be called after wake up and after that one can read the received data.

**Reply by jgromes** [reply]
@Griffon1412 the library you linked to has the same limitation as RadioLib - after waking up, it resets the SX126x device, as can be seen here: https://github.com/beegee-tokyo/SX126x-Arduino/blob/29202d7d19503592fdb20a762e5ae4244c145dd9/src/boards/sx126x/sx126x-board.cpp#L49

The `SX126xIoInit` function is called by `lora_hardware_re_init()`, and resets the radio. The equivalent in RadioLib is to call `radio.begin()` after waking up from deep sleep - but this is something the OP explicitely tries to avoid, as per the title of this discussion.

**Reply by Griffon1412** [reply]
As already said I'm not an expert. All I can say the provided deep sleep example works off the shelf and flawless.  
I was able to add some lines of code to transmit back the received data that woke up the ESP32 from deep sleep.

Maybe this info is interesting for you: [https://github.com/beegee-tokyo/SX126x-Arduino?tab=readme-ov-file#initialize-the-radio-after-cpu-woke-up-from-deep-sleep](url) 
So I suppose by using this functions one does not reset the SX126x.
The link you added points to the normal init , not the re-init after a wake up from deep sleep.

**Reply by DarkZeros** [reply]
> radio-after-cpu-woke-up-from-deep-sleep

In my opinion the example provided is a "hacked" on top solution, which we already discussed in this thread and me personally I am trying to avoid it.
The RadioLib works for many devices, so supporting DeepSleep for ESP32 in particular is a small niche feature.

I would recommend trying the example I posted before, which works with upstream RadioLib, which is basically:
setup():
1. Define RTC_DATA_ATTR std::optional<RadioLib classes (spi/hal/module/radio)>
2. Initialize them once only, checking the optional, in setup(), and creating with proper values.
3. Reinitialize the 2 GPIO pins unconditionally

loop():
4. Use the radio normally, if the wakeup is from the EXT1, then a packet was received, read the radio, etc...
5. Set the EXT1 wakeup in the ESP32 before going to sleep! This way the ESP will wakeup when the radio has some event.

**Reply by cleishm** [reply]
> The `SX126xIoInit` function is called by `lora_hardware_re_init()`, and resets the radio. The equivalent in RadioLib is to call `radio.begin()` after waking up from deep sleep - but this is something the OP explicitely tries to avoid, as per the title of this discussion.

@jgromes: I looked at the library in question and it seems `lora_hardware_re_init()` actually calls [*`SX126xIoReInit()`*](https://github.com/beegee-tokyo/SX126x-Arduino/blob/29202d7d19503592fdb20a762e5ae4244c145dd9/src/boards/sx126x/sx126x-board.cpp#L82) which is essentially identical to the `SX126xIoInit()` method, but doesn't do the reset at the end.

Based on a very superficial review, I think the quality of RadioLib is significantly higher, but it does seem that restoring after deep sleep is a missing function. Would you consider a feature request to add a `reinit()` style method? I understand that there is a lot of state that cannot be retrieved from the module during reset, like spreading factor, but I imagine that is known by the user of the library and could be supplied. Also, perhaps I should open a feature request issue and discuss there?

**Reply by jgromes** [reply]
@cleishm Based on the fact that this request keeps coming back, I think it is worth investigating a little. A fresh issue will be better to discuss the interface and implementation, so feel free to open one.

**Reply by cleishm** [reply]
I created a PR indeed. It turns out that all the initialization done in ::begin can be repeated, except the reset, and leaves the packet buffer intact and able to be read. I have it working nicely on an SX1262 device. #1608 includes an example.

**Reply by StevenCellist** [reply]
> A fresh **issue** will be better **to discuss** the interface and implementation, so feel free to open one.

> I created a PR indeed.

I have a whole lot of opinions on this PR and I fail how to see how the word "indeed" aligns with the proposed issue and discussion. Can you please explain how you interpretated @jgromes' message as an invitation to send a PR instead? That'll help us avert this mismatch in the future as this has now happened on multiple occasions with different people.

**Reply by Griffon1412** [reply]
> I have a whole lot of opinions on this PR and I fail how to see how the word "indeed" aligns with the proposed issue and discussion. Can you please explain how you interpretated @jgromes' message as an invitation to send a PR instead? That'll help us avert this mismatch in the future as this has now happened on multiple occasions with different people.

How smug is that. Someone tries to move things forward and then he gets a comment like that.  Glad I choose the SX1262-Arduino library.

**Reply by HeadBoffin** [reply]
Perhaps @StevenCellist's response could be rewritten as "How smug is that, Someone with no history on the project, a self-proclaimed non-expert who has no insight in to future developments submits a PR when they are asked to open an issue to discuss". Po-tay-toe, po-tar-toe. Ying and Yang.

If you actually did choose the most excellent SX1262-Arduino library, why did you submit a PR here and why do you care that it's been shelved (note, shelved, not binned) pending liaison with the core developers and other interested parties?

There is some merit in discussion to provide for sleep modes & interrupt triggered receiving, one that fits in to the overall picture & architecture. It would be polite to abide by the clear direction given by the repro owner and the principle LoRa/LoRaWAN contributor.

**Reply by cleishm** [reply]
> How smug is that. Someone tries to move things forward and then he gets a comment like that. Glad I choose the SX1262-Arduino library.

It's ok, @Griffon1412. We don't know why Steven is concerned and so it's best not to make attributions.

@StevenCellist / @HeadBoffin - you are correct, definitely a non-expert here. And that's why I felt it best to learn the code path and understand where the concerns might be before kicking off a new discussion. And when I found that I could make it work with such minimal changes, because everything except the call to reset is idempotent, I thought a PR would help to demonstrate that.

In my opinion, code is discussion. I hope you also look at the PR I made with that lens. I promise not to be offended if you tear it apart and show where it's wrong or could be improved :). Or close it, if it's going entirely the wrong way.

**Reply by StevenCellist** [reply]
> In my opinion, code is discussion.

Ah, that's a style that I haven't heard before. Genuinely intrigued, as I've got students in class that may show the same behaviour, but with generally less capability of reflecting / formulating like this. 
And it's also interesting to see how people differ - only recently we had a PR where someone poured his soul into it and a little discussion beforehand would've helped everyone involved. But then here it's considered a kickstarter. Cool! And difficult to properly manage online only.

**Reply by cleishm** [reply]
Agreed. I've also seen some people invest a lot of time into PRs without checking in first, and then get super frustrated when it's not accepted. This isn't one of those cases - the work to make #1608 was probably less than the time it would have taken to write a message making case for it, and given we're all developers, code is often easier to write and understand than english :)

**Reply by Griffon1412** [reply]
@HeadBoffin, keep your arrogance  and _your insights to future developments_ of this obviously rocket science library 'radiolib' for yourself. The linux kernel seems to be just a tiny piece of software compared to this epic work here. 

Btw, it's me who prefers this **- as you call it** - '_most excellent library_', not @cleishm. SX1262-arduino does it exactly it's job, nothing more.  And, it has no collaborators acting so cocky like you and this wannabe teacher(?).

Over and out.
Unsubscribe

**Reply by jgromes** [reply]
Seeing as the discussion has been quite a bit derailed here, I will go ahead and lock this thread. Reinitialization of SX126x has its own discussion in #1608, for the other modules like SX127x (which was the original topic here ...) we will need a different one.
